cpp/src/arrow/array/diff.cc

```diff
@@ -0,0 +1,737 @@
+}
+
+template <typename ArrayType>
+class NullOrListGenerator {
+ public:
+  explicit NullOrListGenerator(const Array& array)
+      : array_(checked_cast<const ArrayType&>(array)), values_(array_.values()) {}
+
+  using SliceType = Slice<ArrayType>;
+
+  NullOr<SliceType> operator()(int64_t index) const {
+    return array_.IsNull(index)
+               ? NullOr<SliceType>()
+               : NullOr<SliceType>(SliceType{array_.value_offset(index),
+                                             array_.value_length(index), *values_});
+  }
+
+ private:
+  const ArrayType& array_;
+  std::shared_ptr<Array> values_;
+};
+
+template <typename Iterator>
+class QuadraticSpaceMyersDiff {
+ public:
+  // represents an intermediate state in the comparison of two arrays
+  struct EditPoint {
+    Iterator base, target;
+
+    bool operator==(EditPoint other) const {
+      return base == other.base && target == other.target;
+    }
+  };
+
+  QuadraticSpaceMyersDiff(Iterator base_begin, Iterator base_end, Iterator target_begin,
+                          Iterator target_end)
+      : base_begin_(base_begin),
+        base_end_(base_end),
+        target_begin_(target_begin),
+        target_end_(target_end),
+        endpoint_base_({ExtendFrom({base_begin_, target_begin_}).base}),
+        insert_({true}) {
+    if (std::distance(base_begin_, base_end_) ==
+            std::distance(target_begin_, target_end_) &&
+        endpoint_base_[0] == base_end_) {
+      finish_index_ = 0;
+    }
+  }
+
+  // beginning of a range for storing per-edit state in endpoint_base_ and insert_
+  uint64_t begin(uint64_t edit_count) const { return edit_count * (edit_count + 1) / 2; }
+
+  // end of a range for storing per-edit state in endpoint_base_ and insert_
+  uint64_t end(uint64_t edit_count) const { return begin(edit_count + 1); }
+
+  EditPoint GetEditPoint(uint64_t edit_count, uint64_t index) const {
+    DCHECK_GE(index, begin(edit_count));
+    DCHECK_LT(index, end(edit_count));
+    int64_t base_distance = endpoint_base_[index] - base_begin_;
+    auto k = static_cast<int64_t>(edit_count) - 2 * (index - begin(edit_count));
+    return {endpoint_base_[index], target_begin_ + (base_distance - k)};
+  }
+
+  EditPoint DeleteOne(EditPoint p) const {
+    if (p.base != base_end_) {
+      ++p.base;
+    }
+    return p;
+  }
+
+  EditPoint InsertOne(EditPoint p) const {
+    if (p.target != target_end_) {
+      ++p.target;
+    }
+    return p;
+  }
+
+  EditPoint ExtendFrom(EditPoint p) const {
+    for (; p.base != base_end_ && p.target != target_end_; ++p.base, ++p.target) {
+      if (*p.base != *p.target) {
+        break;
+      }
+    }
+    return p;
+  }
+
+  void Next() {
+    endpoint_base_.resize(end(edit_count_ + 1), base_begin_);
+    insert_.resize(end(edit_count_ + 1), false);
+
+    // try deleting from base first
+    for (uint64_t i = begin(edit_count_), i_out = end(edit_count_); i != end(edit_count_);
+         ++i, ++i_out) {
```

Comment:
Let's see... here `i_out` is going from `end(edit_count_)` to `end(edit_count_) * 2 - begin(edit_count_)`, right?